Hw6

de-brasi opened the pull request.

de-brasi committed:
make url unique for table "links"

de-brasi committed:
Fix datatype for Telegram chat id

de-brasi committed:
move package "entities" to package "clients"

de-brasi committed:
using new location of UpdateResponse class (clients.entities instead of entities)

de-brasi committed:
better naming of package - move from "client" to common package "clientS"

de-brasi committed:
fix datatype for tg chat ids

de-brasi committed:
init entities

de-brasi committed:
init

de-brasi committed:
init;
todo: checkout "broken" error handling in JdbcLinkRepository::add and JdbcTelegramChatRepository::add (DataAccessException exception, raised after CAUGHT DataAccessException and printing marking-message)

de-brasi committed:
init

de-brasi committed:
undo excluding "org.springframework.boot.autoconfigure.jdbc.DataSourceAutoConfiguration" autoconfiguration from application.yml spring.autoconfigure.exclude - need only when DataSourceAutoConfiguration is not configured for project

de-brasi committed:
move from "domain" package to core module's package

de-brasi committed:
update path to entities

de-brasi committed:
init; structure source is template in task's description

de-brasi committed:
init; realization required

de-brasi committed:
make JdbcTemplate autowired in constructor

de-brasi committed:
add datasource configuration (url, username, password, driver-class-name);
delete excluding "org.springframework.boot.autoconfigure.jdbc.DataSourceAutoConfiguration" in autoconfigure

de-brasi committed:
add logger

de-brasi committed:
fix datasource url

de-brasi committed:
realization

de-brasi committed:
add logging exception and stacktrace for internalErrorHandler

de-brasi committed:
autowire TgChatService to constructor;
using TgChatService in handlers

de-brasi committed:
remove unnecessary autoconfiguration settings for jdbc datasource

de-brasi committed:
remove spring-boot-starter-jdbc and postgresql dependencies

de-brasi committed:
add spring-boot-starter-jdbc and postgresql dependencies

de-brasi committed:
throw IncorrectRequestException when MalformedURLException caught

de-brasi committed:
remove unnecessary todos

de-brasi committed:
delete @Nullable annotation from add method

de-brasi committed:
init

de-brasi committed:
init; need to fix code (in controller and service) and test "removeInCorrectPairLinkNotExists" - more info in test's todo

de-brasi committed:
init; draft version

de-brasi committed:
init

de-brasi committed:
better naming

de-brasi committed:
better naming for LinksNotExistException (to LinksNotAddedException)

de-brasi committed:
init;
add usage hints in todo!

de-brasi committed:
add usage hints in todo!

de-brasi committed:
init

de-brasi committed:
i think it's done, but not tested yet

de-brasi committed:
done

de-brasi committed:
better; add bugfixes; tested version

de-brasi committed:
add constructor from exception

de-brasi committed:
add methods realization

de-brasi committed:
better sql for getting all chat-link bounds

de-brasi committed:
handleGetLinks realization;
tested via swagger;
todo: need to get records only for specified id

de-brasi committed:
handlePostLinks realization;
tested via swagger;
todo: need to support correct link's id

de-brasi committed:
better naming

de-brasi committed:
discard stub entity

de-brasi committed:
untrackLinkForChat realization

de-brasi committed:
add lombok NoArgsConstructor

de-brasi committed:
TEMPORARY(!) comment test need to

de-brasi committed:
logging via log4j.Logger

de-brasi committed:
move from root to "clients" package

de-brasi committed:
using custom exceptions from package moved to

de-brasi committed:
move to package "domain"

de-brasi committed:
using entities ChatLinkBound, TelegramChat, Link from package "domain"

de-brasi committed:
rename field time-in-milliseconds to timeout-in-milliseconds (in properties file, config and beans)

de-brasi committed:
add base url getter

de-brasi committed:
make getters non static

de-brasi committed:
add fetchUpdate methods from url

de-brasi committed:
make botClient bean

de-brasi committed:
better api

de-brasi committed:
better (cascade deletion, unique pairs in track_info and etc)

de-brasi committed:
add fields for storing last checkout time and last value of "update" (got from api)

de-brasi committed:
change sql for inserting new url (add values for field last_check_time and last_update_time)

de-brasi committed:
using linkRepository for adding urls to links table

de-brasi committed:
realization; need to be tested

de-brasi committed:
better api

de-brasi committed:
add realization; need to be tested

de-brasi committed:
remove updating to LinkUpdater::update

de-brasi committed:
add realization for update method (iterate throw links in argument, check if updated, notify link's subscribers)

de-brasi committed:
add function to check is link supported by service

de-brasi committed:
remove unnecessary base url getter

de-brasi committed:
fix error messaging in retrieveLastActivityDateField

de-brasi committed:
better method "compareAndSetLastUpdateTime" - separate sql query: checkout last_update_time and if it outdated than update;
add method "checkLastUpdateTimeChanged";
add little dockstring to "notifyClientsIfUpdatedTimeChanged" and add calling dates updating methods ("actualizeLastUpdateTimeForLink" and "actualizeCheckingTimeForLink") only after success notifying subscribers via bot client;
add methods "actualizeLastUpdateTimeForLink" and "actualizeCheckingTimeForLink";

de-brasi committed:
delete unnecessary updating in "checkLastUpdateTimeChanged" method

de-brasi committed:
better interface; delete methods which should not be public

de-brasi committed:
using new LinkUpdater interface; delegate fetching links need to recheck to linkUpdater

de-brasi committed:
sync with new LinkUpdater interface

de-brasi committed:
linter fix

de-brasi committed:
linter fix

de-brasi committed:
discard unnecessary todos

de-brasi committed:
linter fix

de-brasi committed:
delete irrelevant tests (because they are tests interaction without database)

de-brasi committed:
change default value of timestamp on field 'last_update_time' when create new record in table 'last_update_time'

de-brasi committed:
delete test property source

de-brasi committed:
try to fix problem with not catching DataAccessException;
before creating new record count record with equal url;
add todo described problem with DataAccessException

de-brasi committed:
delete extra import

de-brasi committed:
try to fix problem with not catching DataAccessException;
before creating new record count record with equal chat_id;
add todo described problem with DataAccessException

de-brasi committed:
linter fix

de-brasi committed:
add method for searching by predicate for task 4

de-brasi committed:
add "search" method for task 4 in hw

de-brasi committed:
add "search" method for task 4 in hw; mark all public methods as transactional

de-brasi committed:
add "search" method for task 4 in hw;

de-brasi committed:
add @Override to "search" method

de-brasi committed:
using new methods JdbcLinkRepository::search for getting outdated links

de-brasi committed:
linter fix

de-brasi committed:
add table 'supported_services';
add foreign key to 'supported_services' in table 'links';
add field 'snapshot' in table 'links';

de-brasi committed:
init

de-brasi committed:
include init-supported-services.sql

de-brasi committed:
init

de-brasi committed:
mark as ExternalServiceClient interface realization

de-brasi committed:
init; draft version

de-brasi committed:
init

de-brasi committed:
change 'getActualUpdateTime' signature and name: (public OffsetDateTime getActualUpdateTime) instead of (public boolean checkUpdateTimeChanged)

de-brasi committed:
add getJSONContent method

de-brasi committed:
add getJSONContent method

de-brasi committed:
extends with getJSONContent method

de-brasi committed:
extends with getServiceNameInDatabase method

de-brasi committed:
add constructor from string

de-brasi committed:
add ExternalServicesObserver to constructor;
support new 'links' structure: add service name and snapshot value when inserting new link;
better code;

de-brasi committed:
better exceptions catching (throw DataBaseInteractingException to calling code when DataAccessException caught)

de-brasi committed:
remove unnecessary todo

de-brasi committed:
using ExternalServicesObserver as source of actual update time;
add logic for updating snapshot value when link's update handled

de-brasi committed:
delete unused code

de-brasi committed:
better naming

de-brasi committed:
realization of methods: getActualSnapshot, getRelativeServiceNameInDatabase

de-brasi committed:
better code - remove repeats

de-brasi committed:
extends with method getChangeDescriptionFromResponseBodies

de-brasi committed:
extends with method getChangeDescriptionFromResponseBodies

de-brasi committed:
extends with method getChangeDescriptionFromResponseBodies (with stub message)

de-brasi committed:
getChangingDescription method realization

de-brasi committed:
better

de-brasi committed:
fix test links (from random letter set with leading https to some stackoverflow question)

de-brasi committed:
linter fix

de-brasi committed:
fix with stackoverflow test links

de-brasi committed:
add command 'start'

de-brasi committed:
add logger and some log messages in commands

de-brasi committed:
better code

de-brasi committed:
set goals in todo

de-brasi committed:
register scrapperClient bean

de-brasi committed:
init

de-brasi committed:
use beans with commands from UsedCommandsConfig instead of using builders from ReadyToUseCommands class

de-brasi committed:
fix base url for scrapperClient

de-brasi committed:
better exceptions logging

de-brasi committed:
send message to clients about updates

de-brasi committed:
better github retrieving name and repo regexp

de-brasi committed:
fix url for ScrapperClient

de-brasi committed:
change field-marker of repo's changes presence (from updated_at to pushed_at)

de-brasi committed:
change scheduler interval and delay to 1 min

de-brasi committed:
fix predicate (for getting not checked links for a long time) naming and fix comparator;
change 'notifyClientsIfUpdatedTimeChanged' method's signature: remove argument for subscribers (getting subs can be heavy and );
better logging when exception occurs;
fix method 'getLinkCheckTime': rename from 'getLinkUpdateTime' and check column 'last_check_time' in method's body;
bugfix in actualizeSnapshot (cast text to json)

de-brasi committed:
move to package 'repositories'

de-brasi committed:
update path for repositories (now from 'repositories' package)

de-brasi committed:
collect jdbc repositories to package 'jdbc'

de-brasi committed:
use jdbc repositories from package 'jdbc'

de-brasi committed:
init

de-brasi committed:
init

de-brasi committed:
init

de-brasi committed:
add spring-boot-starter-data-jpa

de-brasi committed:
init

de-brasi committed:
make snapshot nullable

de-brasi committed:
init

de-brasi committed:
rename

de-brasi committed:
init

de-brasi committed:
init

de-brasi committed:
make transactional

de-brasi committed:
remove

de-brasi committed:
init

de-brasi committed:
mark as repository

de-brasi committed:
delete extra import

de-brasi committed:
catch persistence exception

de-brasi committed:
init

de-brasi committed:
add lombok getters and setters

de-brasi committed:
add method get(String url);
add annotation @Transactional on "add" method

de-brasi committed:
add method get(Long id);

de-brasi committed:
init

de-brasi committed:
add cascade remove for chat

de-brasi committed:
better naming

de-brasi committed:
add save method

de-brasi committed:
init

de-brasi committed:
linter fix

de-brasi committed:
collect test to 'jdbc' package

de-brasi committed:
add qualifier for jdbcTgChatService

de-brasi committed:
add qualifier for jdbcLinkService

de-brasi committed:
init

de-brasi committed:
add settings for jpa

de-brasi committed:
add hibernate

de-brasi committed:
i surrender to hibernate :(
change 'snapshot' column from json to text

de-brasi committed:
i surrender to hibernate :(
change 'snapshot' column from json to text

de-brasi committed:
i surrender to hibernate :(
change 'snapshot' column from json to text

de-brasi committed:
init

de-brasi committed:
init

de-brasi committed:
add databaseAccessType field

de-brasi committed:
delete spring annotations (bean/component/etc)

de-brasi committed:
delete spring annotations (bean/component/etc)

de-brasi committed:
delete spring annotations (bean/component/etc)

de-brasi committed:
delete spring annotations (bean/component/etc)

de-brasi committed:
delete spring annotations (bean/component/etc)

de-brasi committed:
fix type

de-brasi committed:
discard qualifier

de-brasi committed:
discard qualifier

de-brasi committed:
specify database-access-type

de-brasi committed:
add @NotNull to databaseAccessType (and fix NorNull import: from jakarta.validation to jetbrains.annotations)

de-brasi committed:
init

de-brasi committed:
fix field name

de-brasi committed:
linter fix

de-brasi committed:
remove unnecessary hibernate property (dialect, used by default)

de-brasi committed:
better (remove TestPropertySource)

de-brasi committed:
fix datasource duplication

de-brasi committed:
add hibernate.connection.driver_class=org.postgresql.Driver

de-brasi committed:
use wWebMvcTest instead of SpringBootTest

de-brasi committed:
better

de-brasi committed:
use WebMvcTest instead of SpringBootTest

de-brasi committed:
add @Qualifier("scrapperTestClient") for tested client

de-brasi committed:
add JpaSupportedServicesRepository for new JpaLinkService constructor

de-brasi committed:
add ExternalServicesObserver for getting initial snapshots

de-brasi committed:
bugfix; check when link is null (and create if so)

de-brasi committed:
mock getActualSnapshot method for ExternalServicesObserver

edikgoose commented on bot/src/main/java/edu/java/bot/configuration/UsedCommandsConfig.java:
```diff
@@ -0,0 +1,184 @@
+    private final ScrapperClient scrapperClient;
+
+    @Bean("commandStart")
+    TelegramBotCommand commandStart() {
```
Сделай их отдельными классами через @Component пожалуйста)

de-brasi replied:
@edikgoose 
То есть чтобы каждая команда была отдельным классом-наследником от TelegramBotCommand, помеченным аннотацией component и потом их заавтовайрить (ну через механизм autowire, не нашел замены англицизму) в конструктор TelegramBotService?

edikgoose commented on bot/src/main/java/edu/java/bot/configuration/UsedCommandsConfig.java:
```diff
@@ -0,0 +1,184 @@
+    }
+
+    private static void logException(Exception e, String commandName) {
+        LOGGER.info((
```
Для ошибок нужно .error level использовать

edikgoose commented on scrapper/src/main/java/edu/java/configuration/JdbcAccessConfiguration.java:
```diff
@@ -0,0 +1,46 @@
+@ConditionalOnProperty(prefix = "app", name = "database-access-type", havingValue = "jdbc", matchIfMissing = true)
+public class JdbcAccessConfiguration {
+    @Bean
+    public LinkService linkService(JdbcChatLinkBoundRepository linkBoundRepository) {
```
Вот к чему привело неудачное архитектурное решение в тз.
По хорошему должно было быть так, что здесь был бы один бин репозитория, который возвращал jdbcRepository. А все сервисы, подтягивали бы его, так как зависели от интерфейса
Теперь я даже настаиваю на том, чтобы это переделать, ибо выглядит чересчур громоздко

de-brasi replied:
@edikgoose, соре, не понимаю в чем тут проблема. Речь о том что сейчас в коде есть интерфейсы для сервисов (напр LinkService, LinkUpdater, etc.) и по две реализации (jpa and jdbc) для каждого интерфейса сервиса вместо одной, которая использует специфичный репозиторий (то есть единственная реализация интерфейса, например для LinkService, которая использует интерфейс LinkRepository, реализация которого - jdbc или jpa, выбирается в conditional-конфигурации)?
Если да, то тг-чате [спрашивали](https://t.me/c/2056008600/3447) у Александра Бирюкова про это:
![Снимок экрана от 2024-04-23 13-06-41](https://github.com/de-brasi/tinkoff-education-java-backend-tasks-spring/assets/89832786/c26f42c1-cc70-44fd-b9c8-35ecac61912f)
![Снимок экрана от 2024-04-23 13-10-57](https://github.com/de-brasi/tinkoff-education-java-backend-tasks-spring/assets/89832786/fc9f3229-ccfa-4f23-a801-e68cbafb4a7d)
Ну, правда не исключаю что я не правильно понял мысль Александра :)
И если ты считаешь этот аргумент "так себе", то можешь дать свой комментарий, а то я тогда запутался? И получается мне тогда надо переделать все репозитории - склеить их в один для каждого типа? То есть склеить JdbcChatLinkBoundRepository, JdbcLinkRepository и JdbcTelegramChatRepository в, условно, JdbcRepository, который будет давать api для работы со всеми сущностями в БД через jdbc; так же поступить с JPA? Это для реализации этого совета:
> По хорошему должно было быть так, что здесь был бы один бин репозитория, который возвращал jdbcRepository.

**P.S**: пардон что запоздало получилось заняться  фиксами и тебе приходится заново вникать в контекст, надеюсь на этой неделе все разгрести по замечаниям

edikgoose replied:
Ну возможно он прав, что при использовании JPA есть контекст, в котором хранятся сущности хибера, и его логика вытекает на уровень сервисов. Поэтому ладно, оставь это как есть)
Я это видел так: у тебя есть сервис ChatService. Внутри методы бизнес логики, которые каким-то образом вызывают методы репозитория. ChatService зависит только от интерфейса ChatRepository. А реализацию репозитория уже подкладывает сам спринг. И сервису должно быть без разницы какая это реализация, он вызывает методы интерфейса.

edikgoose commented on scrapper/src/main/java/edu/java/domain/repositories/jpa/implementations/JpaLinkRepository.java:
```diff
@@ -0,0 +1,87 @@
+public class JpaLinkRepository {
+
+    @PersistenceContext
+    private EntityManager entityManager;
```
Молодец, что через entityManager сделал, но можешь переделать через JpaRepository. Это самое верхнеуровневое общение с базой

edikgoose reviewed (commented):
Задание сделано, над замечаниями подумай)
8/8

de-brasi committed:
use slf4j annotation instead of LogManager.getLogger()

de-brasi committed:
in logException method use logging level "error" (log.info -> log.error)

de-brasi committed:
remove unnecessary imports

de-brasi committed:
change to interface JpaRepository<Link, Long>

de-brasi committed:
init

de-brasi committed:
add @RequiredArgsConstructor annotation;
move getting snapshot from link repository to this service;
use new JpaRepository interfaces (for TrackInfo and Link entities)

de-brasi committed:
change to JpaRepository interface

de-brasi committed:
remove unnecessary SuppressWarning annotation

de-brasi committed:
use new JpaLinkRepository api

de-brasi committed:
use @Slf4j annotation; remove manually created logger

de-brasi committed:
add JpaTrackInfoRepository and ExternalServicesObserver for updated JpaLinkService constructor

de-brasi committed:
use new JpaTelegramChatRepository api

de-brasi committed:
Better code, remove method with incorrect name

de-brasi committed:
fix

de-brasi committed:
fix

de-brasi committed:
change to JpaRepository

de-brasi committed:
use new JpaTelegramChatRepository api

de-brasi committed:
fix

de-brasi committed:
catch second add

de-brasi committed:
add saveIfNotExists method; remove "add" method

de-brasi committed:
use new linkRepository api; add test for twice adding

de-brasi committed:
better

de-brasi committed:
use jpa

de-brasi committed:
in method "saveByChatId" sql query add ON CONFLICT DO NOTHING

de-brasi committed:
add method saveIfNotExists

de-brasi committed:
fix name of method for getting all records linked to chat with specific chatId

de-brasi committed:
use fixed name of method for getting all records linked to chat with specific chatId

de-brasi committed:
linter fix
